### PR TITLE
Copter: refactoring about control_mode

### PR DESCRIPTION
**Pull request by tatsuy**
This is not functional PR.

control_mode's type was defined as Mode::Number in Copter.
but it's defined as Mode Class in Rover and Plane.

Changes are the following:
1. add mode_number function to Mode class
2. change the type of control_mode from Mode::Number to Mode
3. integrate flightmode with control_mode

I tested it in SITL.
 - changed to all modes with disarming
 - tried to change to non-exist flight mode

### CODE REVIEW

**Comment by peterbarker** [comment]
Why the rename of flight_mode to control_mode?

I've been gradually weeding out uses of the control_mode number in the code, so I'm glad to see more work done here.  But the rename makes this a rather long delta!

**Comment by tatsuy** [comment]
@peterbarker thank you for the comment!
> Why the rename of flight_mode to control_mode?

Because it's used "control_mode" in Plane and Rover.
https://github.com/ArduPilot/ardupilot/blob/master/ArduPlane/Plane.h#L287
https://github.com/ArduPilot/ardupilot/blob/master/APMrover2/Rover.h#L205

But I can rework to use flightmode and remove control_mode if you want.

**Comment by peterbarker** [comment]
> @peterbarker thank you for the comment!
> 
> > Why the rename of flight_mode to control_mode?
> 
> Because it's used "control_mode" in Plane and Rover.
> https://github.com/ArduPilot/ardupilot/blob/master/ArduPlane/Plane.h#L287
> https://github.com/ArduPilot/ardupilot/blob/master/APMrover2/Rover.h#L205
> 
> But I can rework to use flightmode and remove control_mode if you want.

I think that would be good to avoid code churn.  I actually find "flightmode" rather more descriptive/informative that "control_mode" anyway....

Your point about consistency is will taken, 'though.  Perhaps another step down the track might be a rename in some of the vehicles.

**Comment by tatsuy** [comment]
>I think that would be good to avoid code churn. I actually find "flightmode" rather more descriptive/informative that "control_mode"

I agree with you.
I will revert the last commit and rename of control_mode to flightmode after confirmation of Randy.

**Comment by rmackay9** [comment]
I don't mind too much either way but I think it would be good to be consistent across vehicles.

**Comment by tatsuy** [comment]
Which should I do for merging?
I will follow the opinion of the maintainer because it's a matter of preference.
(But my first opinion is this PR.)

**Comment by peterbarker** [comment]
@tatsuy Might bring this one up at the next devcall, I think.

If we do stay with "flightmode", it would only be a temporary thing until we do move to something common between vehicles.

How big is this patch if there is no renaming of `flightmode` to `control_mode`?

I'm *really* happy to see us getting rid of the parallel variables - I just think we can eliminate control_mode and retain *flightmode as a first cut.  We may want to rename to just plane `Mode *mode;`  rather than Mode *control_mode;` is all - depending on how the devcall discussion goes.

**Comment by muramura** [comment]
I used to have control_xxx in flight mode filenames. However, it is now mode_xxx. Flightmode is better as a variable to indicate the flight mode. I think it is better to change control_xxx to flightmode.

**Comment by tatsuy** [comment]
@peterbarker 
I have a version that does not rename. It's a bit smaller than renaming.
https://github.com/ArduPilot/ardupilot/compare/master...tatsuy:pr-control_mode_refactoring2

ok, I wait for the result of devcall.

@muramura 
yes, it's a good point.

**Comment by peterbarker** [comment]
On Wed, 5 Feb 2020, murata,katsutoshi wrote:

> I used to have control_xxx in flight mode filenames. However, it is now mode_xxx. Flightmode is better as a variable to indicate the flight mode. I think
> it is better to change control_xxx to flightmode.

Problem is that we can't be common between vehicles, and Randy pointed out 
we should be as common as we can.

So it shouldn't matter which vehicle you are in, you know that e.g. 
"mode->is_armable()" is what you need.  Or "control_mode->is_armable()".

We do not want to use "flightmode->is_armable()" on Rovers :-)

**Comment by muramura** [comment]
@peterbarker san, @tatsuy san, comments thanks.

https://ardupilot.org/dev/docs/porting.html I use the ardupilot wiki to describe the autopilot controller as a flight controller. pixhawk calls it automatic pilot hardware. I think pilotmode, if plain, boat, submarine, rover will be common.

**Comment by tatsuy** [comment]
@peterbarker 
I finally understood your intentions.
I prepared a minimally modified version.
https://github.com/ArduPilot/ardupilot/compare/master...tatsuy:pr-control_mode_refactoring3

This changes are the followings:
・add mode_number function to Mode class
・use flightmode->mode_number() function instead of control_mode

**Comment by peterbarker** [comment]
> @peterbarker
> I finally understood your intentions.
> I prepared a minimally modified version.
> [master...tatsuy:pr-control_mode_refactoring3](https://github.com/ArduPilot/ardupilot/compare/master...tatsuy:pr-control_mode_refactoring3)
> 
> This changes are the followings:
> ・add mode_number function to Mode class
> ・use flightmode->mode_number() function instead of control_mode

Yes!  That's exactly what I was saying - just not very well!

Thanks!

File: ArduCopter/RC_Channel.cpp [hunk 1]
```diff
@@ -180,7 +180,7 @@ void RC_Channel_Copter::do_aux_function(const aux_func_t ch_option, const aux_sw
             break;
 
         case AUX_FUNC::SAVE_TRIM:
-            if ((ch_flag == HIGH) && (copter.control_mode <= Mode::Number::ACRO) && (copter.channel_throttle->get_control_in() == 0)) {
+            if ((ch_flag == HIGH) && (copter.control_mode <= &copter.mode_acro) && (copter.channel_throttle->get_control_in() == 0)) {
```
Comment (tridge) [inline ArduCopter/RC_Channel.cpp]: pointer comparison? needs to be a method on the mode

Comment (rmackay9) [reply ArduCopter/RC_Channel.cpp]: We should replace, "(copter.control_mode <= &copter.mode_acro)" with a new member on the Mode class.  So we should do this:

1. add a new method to the Mode class called something like, "bool autotune_supporte() const { return false; }"
2. acro and stabilize modes should implement this new method but return true.
3. use this new method in the case AUX_FUNC::SAVE_TIME::

Comment (tatsuy) [reply ArduCopter/RC_Channel.cpp]: Thank you for the review.
I fixed it using allows_xxx().

File: ArduCopter/mode_flip.cpp [hunk 2]
```diff
@@ -35,16 +35,13 @@
 // flip_init - initialise flip controller
 bool ModeFlip::init(bool ignore_checks)
 {
-    // only allow flip from ACRO, Stabilize, AltHold or Drift flight modes
-    if (copter.control_mode != Mode::Number::ACRO &&
-        copter.control_mode != Mode::Number::STABILIZE &&
-        copter.control_mode != Mode::Number::ALT_HOLD &&
-        copter.control_mode != Mode::Number::FLOWHOLD) {
+    // only allow flip from ACRO, Stabilize, AltHold or FlowHold flight modes
```
Comment (peterbarker) [inline ArduCopter/mode_flip.cpp]: nitpick: Should make this comment generic - "some flight modes, for example ...".

Comment (tatsuy) [reply ArduCopter/mode_flip.cpp]: Thank you, fixed it.

**Comment by tatsuy** [comment]
@peterbarker 
OK, I did squashing the "rename from" thing.

**Comment by tatsuy** [comment]
Ahh, that's what you mean.
Thanks a lot.

**Comment by tatsuy** [comment]
I rebased it off master.

File: ArduCopter/avoidance_adsb.cpp [hunk 3]
```diff
@@ -20,15 +20,15 @@ MAV_COLLISION_ACTION AP_Avoidance_Copter::handle_avoidance(const AP_Avoidance::O
         copter.failsafe.adsb = true;
         failsafe_state_change = true;
         // record flight mode in case it's required for the recovery
-        prev_control_mode = copter.control_mode;
+        prev_control_mode = copter.flightmode->mode_number();
     }
 
     // take no action in some flight modes
-    if (copter.control_mode == Mode::Number::LAND ||
```
Comment (peterbarker) [inline ArduCopter/avoidance_adsb.cpp]: This is also not your bug, but this should be asking if the vehicle is landing for any reason (`is_landing`) - we could be doing an auto-mode landing and we ought not stop doing that.

Perhaps a future PR...

File: ArduCopter/mode.cpp [hunk 4]
```diff
@@ -258,7 +258,7 @@ bool Copter::set_mode(Mode::Number mode, ModeReason reason)
 
 #if MODE_ZIGZAG_ENABLED == ENABLED && SPRAYER_ENABLED == ENABLED
```
Comment (peterbarker) [inline ArduCopter/mode.cpp]: ... and this should be in `exit_mode`, preferably in the zigzag cpp file... future....

**Comment by tatsuy** [comment]
@peterbarker OK, thank you for your confirmation.

**Comment by rmackay9** [comment]
As discussed on last week's dev call we're delaying any structural changes until early April so that we can finish off more of the Copter-4.0 issues.

This PR has been added to this issue for tracking: https://github.com/ArduPilot/ardupilot/issues/13773

**Comment by tatsuy** [comment]
OK, I got it.

**Comment by peterbarker** [comment]
@tatsuy gah!  Conflicts!  I do like the change - are you interested in rebasing it?

**Comment by tatsuy** [comment]
@peterbarker Sure!

**Comment by peterbarker** [comment]
@tatsuy legit CI failure for revo-mini build:

```
../../ArduCopter/mode_avoid_adsb.cpp: In member function 'bool ModeAvoidADSB::set_velocity(const Vector3f&)':
../../ArduCopter/mode_avoid_adsb.cpp:22:38: error: 'class Copter' has no member named 'mode_avoid_adsb'; did you mean 'mode_guided'?
     if (copter.flightmode != &copter.mode_avoid_adsb) {
```

**Comment by tatsuy** [comment]
I found that ADSB_ENABLE is false in revo-mini.
I think this will solve the problem.

**Comment by tatsuy** [comment]
I changed to use Mode::Number::XXX instead of &copter.mode_xxx because we need to add #if if the mode is disabled.
and rebased.

**Review by peterbarker — APPROVED** [review]
LGTM

**Comment by peterbarker** [comment]
Rebased this to give it a fresh testing

File: ArduCopter/Copter.cpp [hunk 5]
```diff
@@ -650,7 +650,6 @@ bool Copter::get_wp_crosstrack_error_m(float &xtrack_error) const
 Copter::Copter(void)
     : logger(g.log_bitmask),
     flight_modes(&g.flight_mode1),
-    control_mode(Mode::Number::STABILIZE),
```
Comment (khancyr) [inline ArduCopter/Copter.cpp:653]: I not sure that this removal don't have side effect. I will check that.

File: ArduCopter/mode.h [hunk 6]
```diff
@@ -921,6 +945,7 @@ class ModeLoiter : public Mode {
     bool allows_arming(bool from_gcs) const override { return true; };
     bool is_autopilot() const override { return false; }
     bool has_user_takeoff(bool must_navigate) const override { return true; }
+    bool allows_autotune() const override { return true; }
```
Comment (khancyr) [inline ArduCopter/mode.h:949]: don't't loiter and poshold allow flip ?

**Review by khancyr — COMMENTED** [review]
Definitively a good improvement. 
I need to check for the default mode that this doesn't change the default behavior as previously the constructor was ensuring that we will be in STABILIZE

**Comment by rmackay9** [comment]
I'm generally happy with this as long as it doesn't conflict with SCurves.

File: ArduCopter/Copter.h [hunk 7]
```diff
@@ -806,7 +806,7 @@ class Copter : public AP_Vehicle {
     // mode.cpp
     bool set_mode(Mode::Number mode, ModeReason reason);
     bool set_mode(const uint8_t new_mode, const ModeReason reason) override;
-    uint8_t get_mode() const override { return (uint8_t)control_mode; }
```
Comment (rmackay9) [inline ArduCopter/Copter.h:809]: I wonder if we need to check for a nullptr?

Comment (peterbarker) [reply ArduCopter/Copter.h:809]: Within a non-static member function?  I'd hope not :-)

**Comment by peterbarker** [comment]
> Definitively a good improvement.
> I need to check for the default mode that this doesn't change the default behavior as previously the constructor was ensuring that we will be in STABILIZE

It still does :-)

**Comment by peterbarker** [comment]
> I'm generally happy with this as long as it doesn't conflict with SCurves.

Rebased cleanly onto your "-oa3" branch which I thought was the best guess as your current work there.

**Review by peterbarker — APPROVED** [review]
I re-added the #ifs for the throw-mode thing and force-pushed this up again.

I intend to merge it after CI has passed.

File: ArduCopter/mode_flip.cpp [hunk 8]
```diff
@@ -35,16 +35,13 @@
...
-        copter.control_mode != Mode::Number::ALT_HOLD &&
-        copter.control_mode != Mode::Number::FLOWHOLD) {
+    // only allow flip from some flight modes, for example ACRO, Stabilize, AltHold or FlowHold flight modes
+    if (!copter.flightmode->allows_flip()) {
         return false;
     }
 
     // if in acro or stabilize ensure throttle is above zero
-    if (copter.ap.throttle_zero && (copter.control_mode == Mode::Number::ACRO || copter.control_mode == Mode::Number::STABILIZE)) {
+    if (copter.ap.throttle_zero && (copter.flightmode->mode_number() == Mode::Number::ACRO || copter.flightmode->mode_number() == Mode::Number::STABILIZE)) {
```
Comment (rmackay9) [inline ArduCopter/mode_flip.cpp:44]: at some point we should change this check to use "has_manual_throttle()".

File: ArduCopter/Copter.h [hunk 9]
```diff
@@ -806,7 +806,7 @@ class Copter : public AP_Vehicle {
...
+    uint8_t get_mode() const override { return (uint8_t)flightmode->mode_number(); }
```
Comment (rmackay9) [inline ArduCopter/Copter.h:809]: Not for this PR but I think we can delete get_mode() can't we?  I don't think it's used.

**Review by rmackay9 — APPROVED** [review]

**Comment by peterbarker** [comment]
On Mon, 22 Mar 2021, Randy Mackay wrote:

> Not for this PR but I think we can delete get_mode() can't we? I don't think
> it's used.

Pure-virtual, inheritted from AP_Vehicle; used for scripting.

**Comment by rmackay9** [comment]
We can merge this any time.  It's the IntelligentEnergy test that seems to be suddenly failing all the time..

**Comment by peterbarker** [comment]
> We can merge this any time. It's the IntelligentEnergy test that seems to be suddenly failing all the time..

I've merged a fix for that.

**Comment by peterbarker** [comment]
Merged, thanks for your patience @tatsuy

**Pull request merged**